### PR TITLE
Adds Set function to allow jumps longer than 290 years and simplify usage.

### PR DESCRIPTION
**Pull request by yorik**
With Advance function you can't jump more than 290 years (required for some tests).
Also in many places I have to use something like this:
```
fc.Advance(newTime.Sub(fc.Now()))
```
Which could be improved with Set function.

### CODE REVIEW

File: clockwork.go [hunk 1]
```diff
@@ -263,25 +268,49 @@ func (fc *fakeClock) NewTicker(d time.Duration) Ticker {
 	return ft
 }
 
+// notifySleepers finds and notifies all the sleepers waiting until time t.
+func notifySleepers(sleepers []*sleeper, t time.Time) []*sleeper {
```
Comment (DPJacques) [inline clockwork.go:272]: NIT: Can you move this up underneath `notifyBlockers`?

Comment (yorik) [reply clockwork.go]: Done.

File: clockwork.go [hunk 2]
```diff
@@ -263,25 +268,49 @@ func (fc *fakeClock) NewTicker(d time.Duration) Ticker {
...
-			s.awaken(end)
-		} else {
-			newSleepers = append(newSleepers, s)
-		}
-	}
-	fc.sleepers = newSleepers
+
+	fc.sleepers = notifySleepers(fc.sleepers, end)
 	fc.blockers = notifyBlockers(fc.blockers, len(fc.sleepers))
 	fc.time = end
```
Comment (DPJacques) [inline clockwork.go:293]: I'd like to have Advance and Set use the same function under the hood if possible.

How about the following:
```
// Advance advances fakeClock to a new point in time, ensuring channels from any
// previous invocations of After are notified appropriately before returning
func (fc *fakeClock) Advance(d time.Duration) {
	fc.l.Lock()
	defer fc.l.Unlock()
        fc.set(fc.time.Add(d))
}

// Set sets fakeClock to a new point in time, ensuring channels from any
// previous invocations of After are notified appropriately before returning
func (fc *fakeClock) Set(t time.Time) {
	fc.l.Lock()
	defer fc.l.Unlock()
        fc.set(s)
}

// set sets the fakeClock and notifies sleepers and blockers before returning.
// The caller must hold fc.l for the duration.
func (fc *fakeClock) set(t time.Time) {
	fc.sleepers = notifySleepers(fc.sleepers, t)
	fc.blockers = notifyBlockers(fc.blockers, len(fc.sleepers))
	fc.time = t
}
```

File: clockwork.go [hunk 3]
```diff
@@ -263,25 +268,49 @@ func (fc *fakeClock) NewTicker(d time.Duration) Ticker {
...
 
+// Set sets the FakeClock to a new point in time.
+// if it's advance then the function ensures that any existing sleepers are
+// notified appropriately before returning. Otherwise no notification will be
+// done.
+func (fc *fakeClock) Set(t time.Time) {
+	fc.l.Lock()
+	defer fc.l.Unlock()
+
+	if fc.time.After(t) {
```
Comment (DPJacques) [inline clockwork.go:304]: I think it is ok to go through the notification process anyway. It should be a non-op. Advance actually lets users advance backwards using a negative duration and in that case it does go through the notification cycle. I'd prefer not to change that behavior.

File: clockwork.go (hunk 2 above)
Comment (yorik) [reply clockwork.go]: Changed the code to be like that. I think that having separate `notifySleepers` isn't necessary anymore, but it makes the  code nicer.

File: clockwork.go (hunk 3 above)
Comment (yorik) [reply clockwork.go]: You're right. Removed.

**Comment by yorik** [comment]
> I have some comments. I think they should be visible. My github-fu is a little weak.

They are. For me it's the first PR...

**Pull request merged**